cmd/hstspreload/main.go

```diff
@@ -0,0 +1,203 @@
+		os.Exit(0)
+
+	default:
+		fmt.Printf("Unknown command: %s\n", args[0])
+		os.Exit(3)
+	}
+
+	// Wrap this in a function to (statically) enforce a return code.
+	showResult := func() int {
+		if header != nil {
+			fmt.Printf("Observed header: %s%s%s\n", bold, *header, reset)
+		}
+
+		fmt.Printf("\n")
+		switch {
+		case len(issues.Errors) > 0:
+			return 1
+
+		case len(issues.Warnings) > 0:
+			return 2
+
+		default:
+			fmt.Printf("%sSatisfies requirements.%s\n\n", green, reset)
+			return 0
+		}
+	}
+	exitCode := showResult()
+
+	printList(issues.Errors, "Error", red)
+	printList(issues.Warnings, "Warning", yellow)
+
+	os.Exit(exitCode)
+}
+
+func preloadableHeader(header string) (issues hstspreload.Issues) {
+	warnIfNotHeader(header)
+
+	fmt.Printf(
+		"Checking header \"%s%s%s\" for preload requirements...\n",
+		bold, header, reset)
+
+	return hstspreload.PreloadableHeaderString(header)
+}
+
+func removableHeader(header string) (issues hstspreload.Issues) {
+	warnIfNotHeader(header)
+
+	fmt.Printf(
+		"Checking header \"%s%s%s\" for removal requirements...\n",
+		bold, header, reset)
+
+	return hstspreload.RemovableHeaderString(header)
+}
+
+func preloadableDomain(domain string) (header *string, issues hstspreload.Issues) {
+	mustBeDomain(domain)
+
+	fmt.Printf(
+		"Checking domain %s%s%s for preload requirements...\n",
+		underline, domain, reset)
+
+	return hstspreload.PreloadableDomain(domain)
+}
+
+func removableDomain(domain string) (header *string, issues hstspreload.Issues) {
+	mustBeDomain(domain)
+
+	fmt.Printf(
+		"Checking domain %s%s%s for removal requirements...\n",
+		underline, domain, reset)
+
+	return hstspreload.RemovableDomain(domain)
+}
+
+func warnIfNotHeader(str string) {
+	if probablyURL(str) {
+		fmt.Fprintf(os.Stderr,
+			"Warning: please supply an HSTS header string (it appears you supplied a URL).\n")
+	}
+	if probablyDomain(str) {
+		fmt.Fprintf(os.Stderr,
+			"Warning: please supply an HSTS header string (it appears you supplied a domain).\n")
+	}
+}
+
+func mustBeDomain(str string) {
+	if probablyHeader(str) {
+		fmt.Fprintf(os.Stderr,
+			"Invalid argument: please supply a domain (example.com), not a header string.\n")
+		os.Exit(3)
+	}
+
+	if probablyURL(str) {
+		fmt.Fprintf(os.Stderr,
+			"Invalid argument: please supply a domain (example.com) rather than a URL (https://example.com/index.html).\n")
+		os.Exit(3)
+	}
+}
+
+func probablyHeader(str string) bool {
+	return strings.Contains(str, ";") || strings.Contains(str, " ")
+}
+
+func probablyURL(str string) bool {
+	return strings.HasPrefix(str, "http") || strings.Contains(str, ":") || strings.Contains(str, "/")
+}
+
+func probablyDomain(str string) bool {
+	return strings.Contains(str, ".") && !strings.Contains(str, " ")
+}
+
+func printList(list []hstspreload.Issue, title string, fs string) {
+	if len(list) == 0 {
+		return
+	}
+
+	titlePluralized := title
+	if len(list) != 1 {
+		titlePluralized += "s"
```

Comment:
You're right; I'll do that.
(It's weird how I still have an instinctive aversion to modifying parameter variables even if it's totally safe in every language I use on a regular basis unless parameters are explicitly passed by pointer/reference...)